### ISSUE TITLE
lots fo duplicated code in learn.py

### ISSUE THREAD

**Issue by IARI** [Vampouille/superboucle#29]
for instance look at the processNote method - why is it that way?

**Comment by Vampouille** [comment]
I think it's more readable this way. But I agree with you, this is not good... There is also a bug with knowCtrl and knownBtn list : When you overwrite some mapping, old values are not removed from lists of used btn an ctrl. So you cannot switch two controls (for example play and pause).
Those lines :

    self.knownCtrl.add(ctrl_key)
    self.knownBtn.add(btn_key)

are not present in every if part. Maybe we can replace string in self.send_midi_to by a callable function. This way, we can create separate little functions that will handle midi event.

**Comment by IARI** [comment]
That's a good idea - i'll put it in our fork :)

**Comment by IARI** [comment]
Actually I in gui.py there is a similar situation in processNote:
There is a big if construct to decode what function to use for what message.
i would consistently replace it with a mapping from midi-messages (or your abstract representations of them) directly to functions. Would you agree?

**Comment by Vampouille** [comment]
yes, but can you make this modification in a new branch created from last tag 1.2.0 ?

**Comment by IARI** [comment]
Damn i have already started - To be honest i have rewritten the whole device and most of learn.py.
in the main branch of our fork..but I will try to put it in a new branch!

**Comment by IARI** [comment]
Ok, brace yourself @Vampouille, here it is: 660dd953ac96a5a0a9492f65db1e5f281f13242b
